app/models/package_manager/base/mapping_builder.rb

```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class Base
+    # This builder class only exists to ensure we have a defined set
+    # of keys that we allow PackageManager::Base.mapping() to return.
+    class MappingBuilder
+      MISSING = Object.new
+
+      def self.build_hash(name:, description:, repository_url:, homepage: MISSING, keywords_array: MISSING, licenses: MISSING, versions: MISSING)
+        hash = {
+          name: name,
+          description: description,
+          repository_url: repository_url,
+        }
+
+        hash[:homepage] = homepage if homepage != MISSING
+        hash[:keywords_array] = keywords_array if keywords_array != MISSING
+        hash[:licenses] = licenses if licenses != MISSING
+        hash[:versions] = versions if versions != MISSING
+
+        hash
```

Comment:
Can we use `nil` / `Object#not_nil?` instead? Same for VersionBuilder?

Reply:
if we set `nil` as the default value for the kwargs above, then we don't have a way to tell if the kwarg was passed as a param or if the value is just the `nil` default value. We'd only want to set the optional keys above (like `:homepage`) if the user explicitly passed a kwarg for it.

I've run into this problem before in Ruby and haven't found a way to tell if the arg was the default value or not, so this seemed like a clever workaround 😬

Reply:
(maybe there's a better name for this than "MISSING" too)

Reply:
Do we rely on the difference in semantics of a key being present and mapped to `nil` and the key being absent from the hash?

Reply:
we're passing almost the entire `mapped_project()` result to the ActiveRecord objects in `ensure_project`, so it's possible we might set an attribute to `nil` and overwrite data, even though the `mapped_project()` method didn't set that key itself.

the goal with that is mostly to maintain the same functionality as before. Before this, each PackageManager class mapped its own combination of keys to set on Project/Versions. But if I start passing all keys for each PackageManager then we've changed the behavior. I could leave a comment here maybe, e.g. "# TODO: if we determine it's safe to default to 'nil' for any of these keys then we can change the kwarg default to 'nil'." 

(I thought we did with the repository_url [here](https://github.com/librariesio/libraries.io/blob/main/app/models/package_manager/base.rb#L103C1-L103C108), but it turns out it's using `blank?` instead of `key?`, so this wouldn't help there.)